# Richelia_Comparative_Genomics
1. GC content: Includes a script that calculates GC content in coding sequences versus short and long intergenic regions, as well as an R script for plotting Figure 3.
2. GTDB-tk tree: A tree in Newick format used to generate Figure 1 in the manuscript.
3. Genome stats: Contains all scripts used to calculate genome statistics (coding regions, intergenic spacers, transposases, pseudogenes, insertion sequences) along with an R script to plot Figure 2 in the manuscript.
4. Genomes: Contains all the FASTA files of genomes used in this study
5. KO_analysis: Containes KO identifiers for each genome of Richelia and other cyanobacteria analysed in this study. Adittionaly, R script used to check different genes annotated in genomes and to produce Figure 5 in the manuscript.
6. Prokka: Annotated genomes using Prokka.
7. Pseudofinder: Annotated genomes using Pseudofinder.
8. Pseudogenes: Scirpts used to analyse pseudogenes.
9. Richelia Pangenomics: Contains scripts with all the commands used in Anvi'o to generate the pangenome, as well as all the files produced by the Anvi'o pangenomics workflow. Also includes a table with all annotations performed within Anvi'o. 
