---
title: "KO analysis"
author: "Vesna"
date: "2024-04-17"
output: html_document
---

```{r}
library(tidyr)
library(tidyverse)
library(dplyr)
library(stringr)
library(data.table)
library(ggplot2)
library(gridExtra)
#devtools::install_github("david-barnett/microViz")
library(microViz)
```



```{r}
RichPangenome <- read_delim("Richelia_gene_clusters_summary.txt.gz", "\t")
RichPangenome <- RichPangenome %>%
  select(-functional_homogeneity_index, -geometric_homogeneity_index, -combined_homogeneity_index, -aa_sequence)

#In the new variable "core" we define "core" as "Core", 'Singletons' and 'Singletons' and 

```

#Make a new table with adittion of 'Eggnog_KEGG_KO' and 'COG20_FUNCTIONS', 'KEGG_BRITE' and 'KEGG_Class'


```{r}
# Selecting columns to create a new data frame

GCbyKO_Richelia <- RichPangenome %>%
  # Replace NA in EGGNOG_KEGG_KO with "Unclassified"
  mutate(EGGNOG_KEGG_KO = replace_na(EGGNOG_KEGG_KO, "Unclassified")) %>%
  # Label rows with more than one KO as "Ambiguous"
  mutate(EGGNOG_KEGG_KO = if_else(str_detect(EGGNOG_KEGG_KO, ","), "Ambiguous", EGGNOG_KEGG_KO)) %>%   group_by(bin_name, genome_name, num_genes_in_gene_cluster, EGGNOG_KEGG_KO, KEGG_BRITE, KEGG_Class) %>%
  summarise(num_corrected_genes = sum(1 / num_genes_in_gene_cluster), .groups = "drop")


```


#Add column symbiosis_location

```{r}

GCbyKO_Richelia <- GCbyKO_Richelia %>%
  mutate(symbiosis_location = case_when(
    genome_name %in% c("SAMN33830777","SAMN33830812","SAMN33830791", "GCF_900185595_1_CalSC01_2013_genomic") ~ "epibionts",
    genome_name %in% c("GCA_000613065_1_RintRC_1_genomic", "GCA_013214565_1_ASM1321456v1_genomic") ~ "partial_endobionts",
    TRUE ~ "endobionts"
  ))

#Rename genome names to shorter names used throughout the manuscript and set the sepcific order of the genomes
GCbyKO_Richelia$genome_name <- recode_factor(GCbyKO_Richelia$genome_name, 
  "SAMN33830777" = "MO_167.B12",
  "SAMN33830812" = "MO_192.B10",
  "SAMN33830791" =  "MO_167.B42",
  "GCF_900185595_1_CalSC01_2013_genomic" = "RrhiSC01",
  "TARA_PON_109_MAG_00086" = "TARA_PON",
  "GCA_000350125_1_ASM35012v1_genomic" = "ReuHM01",
  "TARA_MED_95_MAG_00146" = "TARA_MED",
  "BGEO_SAMN07136523_METAG_IKJKMCAK" = "BGEO_MAG",
  "GCA_002470035_1_ASM247003v1_genomic" = "UBA7409",
  "GCA_002377925_1_ASM237792v1_genomic" = "UBA3481",
  "GCA_900299445_1_Richelia_intracellularis_AM_2804_genomic" = "AM_2804",
  "GCA_000350105_1_ASM35010v1_genomic" = "ReuHH01",
  "GCA_000613065_1_RintRC_1_genomic" = "RintRC01",
  "GCA_013214565_1_ASM1321456v1_genomic" = "MAG_DT_104",
  .ordered = TRUE)


```







```{r}
# Filter entries where EGGNOG_KEGG_KO starts with 'ko:'
filtered_data <- GCbyKO_Richelia %>%
  filter(str_starts(EGGNOG_KEGG_KO, "ko:"))

# Extract the KO number and count occurrences for each genome
ko_counts <- filtered_data %>%
  mutate(KO_Number = str_extract(EGGNOG_KEGG_KO, "(?<=ko:)[K0-9]+")) %>%
  group_by(genome_name, KO_Number) %>%
  summarise(Count = n(), .groups = 'drop') %>%
  group_by(genome_name) %>%
  summarise(Total_KO_Counts = sum(Count))

# View the result
print(ko_counts)


# Extract the KO number and count occurrences for each genome and each bin
ko_counts_per_genome_bin <- filtered_data %>%
  mutate(KO_Number = str_extract(EGGNOG_KEGG_KO, "(?<=ko:)[K0-9]+")) %>%
  group_by(genome_name, bin_name, KO_Number) %>%
  summarise(Count = n(), .groups = 'drop') %>%
  group_by(genome_name, bin_name) %>%
  summarise(Total_KO_Counts = sum(Count), .groups = 'drop')

# View the result
print(ko_counts_per_genome_bin)
```



```{r}
# Define the categories of interest
categories <- c("Energy metabolism", "Metabolism of cofactors and vitamins", "Carbohydrate metabolism",
                "Amino acid metabolism", "Translation", "Membrane transport", "Nucleotide metabolism",
                "Signal transduction", "Replication and repair", "Lipid metabolism",
                "Metabolism of terpenoids and polyketides", "Folding, sorting and degradation",
                "Metabolism of other amino acids", "Xenobiotics biodegradation and metabolism",
                "Glycan biosynthesis and metabolism", "Biosynthesis of other secondary metabolites",
                "Cell growth and death", "Transcription", "Transport and catabolism", "Cell motility")

# Adjusted function to never return NULL and handle outputs uniformly
clean_kegg_brite <- function(brite_string) {
  if (is.na(brite_string)) {
    return(NA)
  } else {
    # Use str_detect to identify categories present in the string
    matched_categories <- lapply(categories, function(cat) {
      if (str_detect(brite_string, fixed(cat))) {
        return(cat)
      } else {
        return(character(0))  # return empty character vector instead of NULL
      }
    })
    # Flatten the list and remove empty entries
    matched_categories <- unlist(matched_categories)
    if (length(matched_categories) == 0) {
      return(NA)  # Return NA if no categories match
    } else {
      return(unique(matched_categories))
    }
  }
}

# Apply the function using lapply for consistent list output
GCbyKO_Richelia$Cleaned_KEGG_BRITE <- lapply(GCbyKO_Richelia$KEGG_BRITE, clean_kegg_brite)

# View some of the cleaned data
head(GCbyKO_Richelia$Cleaned_KEGG_BRITE)

```



```{r}


# Extract KO_Number right after its mutations are done
GCbyKO_Richelia <- GCbyKO_Richelia %>%
  mutate(KO_Number = str_extract(EGGNOG_KEGG_KO, "(?<=ko:)[K0-9]+"))

# Now expand each Cleaned_KEGG_BRITE into its own row
GCbyKO_Richelia_expanded <- GCbyKO_Richelia %>%
  unnest(Cleaned_KEGG_BRITE)

# Step 2: Ensure KO_Number extraction has been done
#GCbyKO_Richelia$KO_Number <- sapply(GCbyKO_Richelia$EGGNOG_KEGG_KO, function(x) str_extract(x, #"(?<=ko:)[K0-9]+"))

# Step 3: Group by genome_name, bin_name, and Cleaned_KEGG_BRITE, then count distinct KO_Numbers
ko_counts_per_category <- GCbyKO_Richelia_expanded %>%
  group_by(genome_name, bin_name, Cleaned_KEGG_BRITE) %>%
  summarise(Total_KO_Counts = n_distinct(KO_Number), .groups = 'drop')

# View the result
print(ko_counts_per_category)


```


```{r}
ko_counts_per_category_location <- GCbyKO_Richelia_expanded %>%
  group_by(symbiosis_location, bin_name, Cleaned_KEGG_BRITE) %>%
  summarise(Total_KO_Counts = n_distinct(KO_Number), .groups = 'drop') %>%
  complete(symbiosis_location, bin_name, Cleaned_KEGG_BRITE, fill = list(Total_KO_Counts = 0))

ko_counts_per_category_location_2 <- GCbyKO_Richelia_expanded %>%
  group_by(symbiosis_location, Cleaned_KEGG_BRITE) %>%
  summarise(Total_KO_Counts = n_distinct(KO_Number), .groups = 'drop') %>%
  complete(symbiosis_location, Cleaned_KEGG_BRITE, fill = list(Total_KO_Counts = 0))

```


```{r}
# Convert the categorical variables to factors if they are not already
ko_counts_per_category$genome_name <- factor(ko_counts_per_category$genome_name)
ko_counts_per_category$bin_name <- factor(ko_counts_per_category$bin_name)
ko_counts_per_category$Cleaned_KEGG_BRITE <- factor(ko_counts_per_category$Cleaned_KEGG_BRITE)

# Reverse the order of genome names
genome_order <- c("BGEO_MAG", "TARA_MED", "UBA7409", "UBA3481",
                  "ReuHH01", "AM_2804", "ReuHM01", "TARA_PON",
                  "MAG_DT_104", "RintRC01", "RrhiSC01", "MO_167.B42", "MO_192.B10", "MO_167.B12")

# Ensure genome_name is a factor and then relevel it
ko_counts_per_category$genome_name <- factor(ko_counts_per_category$genome_name, levels = genome_order)


# Assuming the distinct_palette() function generates a palette and 'kelly' specifies the type
kelly_colors <- distinct_palette(pal = "kelly")
brewerPlus <- distinct_palette()
scales::show_col(brewerPlus)

# Now we can see the colors
scales::show_col(kelly_colors)

#Plot stacked bar plot of K0 counts including NA K0 numbers:
ga <- ggplot(ko_counts_per_category, aes(x = genome_name, y = Total_KO_Counts, fill = Cleaned_KEGG_BRITE)) +
  geom_bar(stat = "identity", position = "stack") +  # Stacked bar plot
  coord_flip() +  # Flip the coordinates for a horizontal bar plot
  scale_fill_manual(values = brewerPlus) +  # Use your defined colors
  theme_minimal() +  # Use a minimal theme
  theme(legend.position = "top",  # Position the legend at the top
        panel.border = element_rect(colour = "black", fill=NA),  # Add a border around the plot
        axis.line = element_line(color = "black"),  # Add axis lines
        panel.grid.major = element_blank(),  # Remove major grid lines
        panel.grid.minor = element_blank(),  # Remove minor grid lines
        axis.text.y = element_text(angle = 0, hjust = 1)) +  # Adjust y-axis text alignment
  labs(x = "Genome Name", y = "Number of K0 IDs", fill = "KEGG Functional Categories")  # Update labels

ga


# Calculate percentages within each genome (without considering bins)
ko_per_category <- ko_counts_per_category %>%
  group_by(genome_name, Cleaned_KEGG_BRITE) %>%
  summarise(Total_KO_Counts = sum(Total_KO_Counts)) %>%
  group_by(genome_name) %>%
  mutate(Total_Per_Genome = sum(Total_KO_Counts)) %>%  
  mutate(Percentage = (Total_KO_Counts / Total_Per_Genome) * 100) %>%  
  ungroup()



# Check if percentages sum to 100 within each genome_name group
check_sums <- ko_per_category %>%
  group_by(genome_name) %>%
  summarise(Total_Percentage = sum(Percentage))  # Summaries to check if total percentages equal 100%


# Create the plot
ga_1 <- ggplot(data = ko_per_category, aes(x = genome_name, y = Percentage, fill = Cleaned_KEGG_BRITE)) +
  geom_bar(stat = "identity", position = "fill", color = "grey") +  # Stacked bar plot with black borders
  scale_fill_manual(values = brewerPlus) +  # Use custom colors for the categories
  labs(fill = "KEGG Functional Categories", x = "Genome Name", y = "Percentage of KO IDs") +  # Adjust labels
  theme_minimal() +  # Use a minimal theme
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),  # Rotate x-axis labels
    legend.position = "bottom",  # Position legend at the bottom
    panel.border = element_rect(colour = "black", fill=NA),  # Add a border around the plot
    axis.line = element_line(color = "black"),  # Add axis lines
    panel.grid.major = element_blank(),  # Remove major grid lines
    panel.grid.minor = element_blank(),  # Remove minor grid lines
    axis.text.y = element_text(angle = 0, hjust = 1)  # Adjust y-axis text alignment
  ) +
  scale_y_continuous(labels = scales::percent_format()) +  # Ensure percentages are formatted
  coord_flip() # Flip the coordinates


ga_1



```




```{r}

# Combine the plots vertically
combined_plot <- arrangeGrob(ga, ga_1, ncol = 1)

# Save the combined plot to a PDF file
ggsave("Combined_Plots.pdf", plot = combined_plot, width = 10, height = 12, units = "in")


```




```{r}
# Count the number of unassigned (NA) KO numbers and the number of annotated KO numbers for each genome
ko_annotation_summary <- ko_counts_per_category %>%
  mutate(Annotated = if_else(is.na(Cleaned_KEGG_BRITE), "Unassigned", "Annotated")) %>%
  group_by(genome_name, Annotated) %>%
  summarise(Total_KO_Counts = sum(Total_KO_Counts), .groups = 'drop')

# Spread the summary to have a column for Annotated and Unassigned KO numbers
ko_annotation_summary_wide <- ko_annotation_summary %>%
  pivot_wider(names_from = Annotated, values_from = Total_KO_Counts, values_fill = list(Total_KO_Counts = 0))

# Ensure genome_name is a factor with the desired order
ko_annotation_summary_wide <- ko_annotation_summary_wide %>%
  mutate(genome_name = factor(genome_name, levels = genome_order))

# Calculate the total KO numbers for each genome
ko_annotation_summary_wide <- ko_annotation_summary_wide %>%
  mutate(Total_KO_Counts = Annotated + Unassigned)

# Calculate percentages for Annotated and Unassigned KO numbers
ko_annotation_summary_wide <- ko_annotation_summary_wide %>%
  mutate(
    Annotated_Percentage = round((Annotated / Total_KO_Counts) * 100, 1),
    Unassigned_Percentage = round((Unassigned / Total_KO_Counts) * 100, 1)
  )

# Display the summary table with counts and percentages, ensuring the correct order
knitr::kable(ko_annotation_summary_wide[order(ko_annotation_summary_wide$genome_name), ])

# Optional: Save the summary table as a TSV file
write.table(ko_annotation_summary_wide, file = "KO_Annotation_Summary_With_Percentages.tsv", sep = "\t", row.names = FALSE, quote = FALSE)

```




```{r}
#Plot stacked bar plot of K0 counts excluding NA K0 numbers:
# Filter out rows where 'Cleaned_KEGG_BRITE' is NA
filtered_data <- ko_counts_per_category %>%
  filter(!is.na(Cleaned_KEGG_BRITE))

# Create your ggplot with the specified colors from the 'kelly' palette
pa <- ggplot(ko_counts_per_category, aes(x = genome_name, y = Total_KO_Counts, fill = Cleaned_KEGG_BRITE)) +
  geom_bar(stat = "identity", position = "stack") +
  facet_wrap(~bin_name, scales = "free_x", nrow = 1) +  # Each facet with its own x-scale and in a single row
  coord_flip() +
  scale_fill_manual(values = brewerPlus) + # Use the kelly palette
  theme_minimal() +
  theme(
    legend.position = "top",  # Match the legend position to `pb`
    strip.background = element_blank(),
    strip.text.x = element_text(face = "bold"),
    panel.border = element_rect(colour = "black", fill = NA),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)  # Rotate X-axis labels to 45 degrees
  ) + 
  labs(x = "Number of KO IDs", y = "Genome Name", fill = "KEGG Functional Categories")  # Match the labels to `pb`

pa

# Create the ggplot with the filtered data
pb <- ggplot(filtered_data, aes(x = genome_name, y = Total_KO_Counts, fill = Cleaned_KEGG_BRITE)) +
  geom_bar(stat = "identity", position = "stack") +
  facet_wrap(~bin_name, scales = "free_x", nrow = 1) +
  coord_flip() +
  scale_fill_manual(values = brewerPlus) + # Use the modified kelly palette
  theme_minimal() +
  theme(
    legend.position = "top",
    strip.background = element_blank(),
    strip.text.x = element_text(face = "bold"),
    panel.border = element_rect(colour = "black", fill = NA),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)  # Rotate X-axis labels to 45 degrees
  ) + 
  labs(x = "Number of KO IDs", y = "Genome Name", fill = "KEGG Functional Categories")

pb


```



```{r}
# Combine the two plots vertically
combined_plot <- pa / pb

# Save the combined plot as a PDF
ggsave("combined_plots_vertical.pdf", plot = combined_plot, device = "pdf", width = 10, height = 10, units = "in")
```








```{r}


# Create a summary table of KO counts per genome and KEGG category
ko_counts_summary <- filtered_data %>%
  group_by(genome_name, Cleaned_KEGG_BRITE) %>%
  summarise(Total_KO_Counts = sum(Total_KO_Counts), .groups = 'drop') %>%
  pivot_wider(names_from = Cleaned_KEGG_BRITE, values_from = Total_KO_Counts, values_fill = list(Total_KO_Counts = 0))

# Display the summary table
knitr::kable(ko_counts_summary)

# Optional: Save the summary table as a TSV file
#write.table(ko_counts_summary, file = "KO_Counts_Summary.tsv", sep = "\t", row.names = FALSE, quote = FALSE)



# Calculate the total KO counts per genome
ko_counts_percentages <- ko_counts_summary %>%
  mutate(Total_KO_Counts = rowSums(select(., -genome_name))) %>%
  mutate(across(-genome_name, ~ round(. / Total_KO_Counts * 100, 1)))

# Remove the Total_KO_Counts column as it is no longer needed for percentage table
ko_counts_percentages <- select(ko_counts_percentages, -Total_KO_Counts)

# Display the percentage table
knitr::kable(ko_counts_percentages)

# Optional: Save the percentage table as a TSV file
#write.table(ko_counts_percentages, file = "KO_Counts_Percentages.tsv", sep = "\t", row.names = FALSE, quote = FALSE)


```




#Now plot the same plots but for external, semi-internal and internal:
```{r}
# Setting the levels for symbiosis_location
ko_counts_per_category_location$symbiosis_location <- factor(
  ko_counts_per_category_location$symbiosis_location,
  levels = c("endobionts", "partial_endobionts", "epibionts")
)

# Plot by symbiosis location only
plot1 <- ggplot(ko_counts_per_category_location, aes(x = symbiosis_location, y = Total_KO_Counts, fill = Cleaned_KEGG_BRITE)) +
  geom_bar(stat = "identity", position = "stack") +
  coord_flip() +
  scale_fill_manual(values = brewerPlus) +
  theme_minimal() +
  theme(legend.position = "bottom",
        panel.border = element_rect(colour = "black", fill=NA),
        axis.line = element_line(color = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = "Symbiosis Location", y = "Total KO Counts", fill = "KEGG Brite Category")

plot1

# Plot by symbiosis location and bin name
plot2 <- ggplot(ko_counts_per_category_location, aes(x = symbiosis_location, y = Total_KO_Counts, fill = Cleaned_KEGG_BRITE)) +
  geom_bar(stat = "identity", position = "stack") +
  facet_wrap(~bin_name, scales = "free_y") +
  coord_flip() +
  scale_fill_manual(values = brewerPlus) +
  theme_minimal() +
  theme(legend.position = "bottom",
        strip.background = element_blank(),
        strip.text.x = element_text(face = "bold"),
        panel.border = element_rect(colour = "black", fill=NA),
        axis.line = element_line(color = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.text.y = element_text(angle = 0, hjust = 1)) +
  labs(x = "Symbiosis Location", y = "Total KO Counts", fill = "KEGG Brite Category")

plot2

```


```{r}
# Filter out NA values from the Cleaned_KEGG_BRITE column
filtered_data <- ko_counts_per_category_location %>%
  filter(!is.na(Cleaned_KEGG_BRITE))

# Now, ensuring that the symbiosis_location is factored correctly
filtered_data$symbiosis_location <- factor(
  filtered_data$symbiosis_location,
  levels = c("endobionts", "partial_endobionts", "epibionts")
)

```


```{r}
# Plot by symbiosis location only
plot1 <- ggplot(filtered_data, aes(x = symbiosis_location, y = Total_KO_Counts, fill = Cleaned_KEGG_BRITE)) +
  geom_bar(stat = "identity", position = "stack") +
  coord_flip() +
  scale_fill_manual(values = brewerPlus) +
  theme_minimal() +
  theme(legend.position = "bottom",
        panel.border = element_rect(colour = "black", fill=NA),
        axis.line = element_line(color = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = "Symbiosis Location", y = "Total KO Counts", fill = "KEGG Brite Category")

# Print the first plot
print(plot1)

# Plot by symbiosis location and bin name
plot2 <- ggplot(filtered_data, aes(x = symbiosis_location, y = Total_KO_Counts, fill = Cleaned_KEGG_BRITE)) +
  geom_bar(stat = "identity", position = "stack") +
  facet_wrap(~bin_name, scales = "free_y") +
  coord_flip() +
  scale_fill_manual(values = brewerPlus) +
  theme_minimal() +
  theme(legend.position = "bottom",
        strip.background = element_blank(),
        strip.text.x = element_text(face = "bold"),
        panel.border = element_rect(colour = "black", fill=NA),
        axis.line = element_line(color = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.text.y = element_text(angle = 0, hjust = 1)) +
  labs(x = "Symbiosis Location", y = "Total KO Counts", fill = "KEGG Brite Category")

# Print the second plot
print(plot2)

```


```{r}

# Filter out rows with NA in 'Cleaned_KEGG_BRITE'
filtered_data <- ko_counts_per_category_location_2 %>%
  filter(!is.na(Cleaned_KEGG_BRITE))

# Setting the order for symbiosis_location
filtered_data$symbiosis_location <- factor(
  filtered_data$symbiosis_location,
  levels = c("endobionts", "partial_endobionts", "epibionts")
)


```



```{r}
# Calculate the total counts per KEGG BRITE category and symbiosis location
ordered_kegg <- filtered_data %>%
  group_by(Cleaned_KEGG_BRITE) %>%
  summarize(Total_Counts = sum(Total_KO_Counts)) %>%
  ungroup() %>%
  arrange(desc(Total_Counts)) %>%
  .$Cleaned_KEGG_BRITE

# Reorder the factor levels based on the calculated total counts
filtered_data$Cleaned_KEGG_BRITE <- factor(filtered_data$Cleaned_KEGG_BRITE, levels = ordered_kegg)

# Assuming 'brewerPlus' has been defined with colors for each symbiosis_location
# brewerPlus <- setNames(brewerPlus, levels(filtered_data$Cleaned_KEGG_BRITE))

# Plot with normal bars for each symbiosis location and KEGG category, in descending order of KO counts
plot1 <- ggplot(filtered_data, aes(x = Cleaned_KEGG_BRITE, y = Total_KO_Counts, fill = symbiosis_location)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  coord_flip() +  # Flip coordinates for a horizontal bar plot
  scale_fill_manual(values = brewerPlus) +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    panel.border = element_rect(colour = "black", fill=NA),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.y = element_text(angle = 0, hjust = 1)
  ) +
  labs(x = "KEGG Brite Category", y = "Total KO Counts", fill = "Symbiosis Location")

# Print the plot
print(plot1)


plot2 <- ggplot(filtered_data, aes(x = Cleaned_KEGG_BRITE, y = Total_KO_Counts, fill = symbiosis_location)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = Total_KO_Counts), 
            position = position_dodge(width = 0.9), 
            vjust = -0.5, 
            size = 3) +  # Adjust size as needed
  coord_flip() +  # Flip coordinates for a horizontal bar plot
  scale_fill_manual(values = brewerPlus) +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    panel.border = element_rect(colour = "black", fill=NA),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.y = element_text(angle = 0, hjust = 1)
  ) +
  labs(x = "KEGG Brite Category", y = "Total KO Counts", fill = "Symbiosis Location")

# Print the plot
print(plot2)



ggsave("functional_cat.pdf", plot = plot1, device = "pdf", width = 10, height = 8, units = "in")

```




```{r}
# Calculate gene loss between endobionts and epibionts, and between partial endobionts and epibionts
wide_data <- wide_data %>%
  mutate(Gene_Loss_Endobionts_vs_Epibionts = epibionts - endobionts,
         Gene_Loss_Partial_Endobionts_vs_Epibionts = epibionts - partial_endobionts)

# Sort by the gene loss (from highest to lowest) for endobionts vs epibionts
top3_loss_endobionts <- wide_data %>%
  arrange(desc(Gene_Loss_Endobionts_vs_Epibionts)) %>%
  slice(1:3)  # Get the top 3 categories

# Print the top 3 categories with the highest gene loss for endobionts vs epibionts
print(top3_loss_endobionts$Cleaned_KEGG_BRITE)

# Sort by the gene loss (from highest to lowest) for partial endobionts vs epibionts
top3_loss_partial_endobionts <- wide_data %>%
  arrange(desc(Gene_Loss_Partial_Endobionts_vs_Epibionts)) %>%
  slice(1:3)  # Get the top 3 categories

# Print the top 3 categories with the highest gene loss for partial endobionts vs epibionts
print(top3_loss_partial_endobionts$Cleaned_KEGG_BRITE)

```





```{r}
# Calculate percentage gene loss
wide_data <- wide_data %>%
  mutate(Percent_Loss_Endobionts_vs_Epibionts = 100 * (epibionts - endobionts) / epibionts,
         Percent_Loss_Partial_Endobionts_vs_Epibionts = 100 * (epibionts - partial_endobionts) / epibionts)

# Sort by the percentage gene loss (from highest to lowest) for endobionts vs epibionts
top3_percent_loss_endobionts <- wide_data %>%
  arrange(desc(Percent_Loss_Endobionts_vs_Epibionts)) %>%
  slice(1:3)  # Get the top 3 categories

# Print the top 3 categories with the highest percentage gene loss for endobionts vs epibionts
print(top3_percent_loss_endobionts$Cleaned_KEGG_BRITE)

# Sort by the percentage gene loss (from highest to lowest) for partial endobionts vs epibionts
top3_percent_loss_partial_endobionts <- wide_data %>%
  arrange(desc(Percent_Loss_Partial_Endobionts_vs_Epibionts)) %>%
  slice(1:3)  # Get the top 3 categories

# Print the top 3 categories with the highest percentage gene loss for partial endobionts vs epibionts
print(top3_percent_loss_partial_endobionts$Cleaned_KEGG_BRITE)

```



```{r}
aggregated_counts <- ko_counts_per_category %>%
  filter(genome_name %in% c("RrhiSC01", "RintRC01", "ReuHH01")) %>%
  filter(!is.na(Cleaned_KEGG_BRITE)) %>%
  group_by(genome_name, Cleaned_KEGG_BRITE) %>%
  summarize(Total_KO_Counts = sum(Total_KO_Counts, na.rm = TRUE)) %>%
  ungroup()

# Order the KEGG BRITE categories by the total counts
aggregated_counts <- aggregated_counts %>%
  mutate(Cleaned_KEGG_BRITE = fct_reorder(Cleaned_KEGG_BRITE, Total_KO_Counts, .desc = TRUE))

# Plot with normal bars for the selected genomes and KEGG category
plot <- ggplot(aggregated_counts, aes(x = Cleaned_KEGG_BRITE, y = Total_KO_Counts, fill = genome_name)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  coord_flip() +  # Flip coordinates for a horizontal bar plot
  scale_fill_manual(values = brewerPlus) +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    panel.border = element_rect(colour = "black", fill=NA),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.y = element_text(angle = 0, hjust = 1)
  ) +
  labs(x = "KEGG Brite Category", y = "Total KO Counts", fill = "Genome Name")
plot

```



#Checking for nitrogene fixation genes in the core

```{r}
# List of nitrogenase K numbers with gene names
nitrogenase_K_numbers <- c(
  # nifH - Nitrogenase reductase
  "ko:K02588",

  # nifD - Nitrogenase molybdenum-iron protein subunit alpha
  "ko:K02586",

  # nifK - Nitrogenase molybdenum-iron protein subunit beta
  "ko:K02591",

  # nifE - Nitrogenase molybdenum-cofactor biosynthesis protein
  "ko:K02587",

  # nifN - Nitrogenase molybdenum-cofactor biosynthesis protein
  "ko:K02592",

  # nifX - Iron-molybdenum cofactor processing protein
  "ko:K02596",

  # nifA - Nif-specific regulatory protein
  "ko:K02584",

  # nifB - FeMo cofactor biosynthesis protein
  "ko:K02585",

  # nifQ - Molybdenum ion binding protein
  "ko:K15790",

  # nifS - Cysteine desulfurase
  "ko:K04487",

  # nifT - Nitrogen fixation protein
  "ko:K02593",

  # nifU - Fe-S cluster assembly protein
  "ko:K04488",

  # nifV - Homocitrate synthase
  "ko:K02594",

  # nifW - Nitrogenase stabilizing/protective protein
  "ko:K02595",

  # nifZ - Iron-sulfur cofactor synthesis protein
  "ko:K02597",

  # nifJ - Pyruvate-flavodoxin oxidoreductase
  "ko:K03737",

  # nifL - Nitrogen fixation negative regulator
  "ko:K23916",

  # nifP - Serine acetyltransferase
  "ko:K00640",
  )

# Check if nitrogenase K numbers are present in the dataset
nitrogenase_genes_present <- RichPangenome %>%
  filter(EGGNOG_KEGG_KO %in% nitrogenase_K_numbers)

# Check if nitrogenase genes are marked as core
core_nitrogenase_genes <- RichPangenome %>%
  filter(bin_name == "Core" & EGGNOG_KEGG_KO %in% nitrogenase_K_numbers)

print(core_nitrogenase_genes)

# Extract core genome genes
core_genes <- RichPangenome %>%
  filter(bin_name == "Core") %>%
  pull(EGGNOG_KEGG_KO)

# Check if all nitrogenase genes are in the core genome
all_in_core <- all(nitrogenase_K_numbers %in% core_genes)

# Output the result for core genes
if (all_in_core) {
  print("All nitrogenase genes are in the core genome.")
} else {
  print("Not all nitrogenase genes are in the core genome.")
}

# Check the unique values in the 'bin_name' column to identify the correct value for core genes
unique_bins <- unique(RichPangenome$bin_name)
print(unique_bins)

# Extract core genome genes
core_genes <- RichPangenome %>%
  filter(bin_name == "Core") %>%
  pull(EGGNOG_KEGG_KO)

# Find common nitrogenase genes in the core genome
common_genes <- intersect(nitrogenase_K_numbers, core_genes)

# Count how many nitrogenase genes are in the core genome
num_common_genes <- length(common_genes)

# Output the result
print(paste(num_common_genes, "out of", length(nitrogenase_K_numbers), "nitrogenase genes are in the core genome."))
print("The nitrogenase genes in the core genome are:")
print(common_genes)


# Check if nitrogenase genes are marked as flexible (Flexible_1, Flexible_2, ...)
flexible_nitrogenase_genes <- RichPangenome %>%
  filter(startsWith(bin_name, "Flexible") & EGGNOG_KEGG_KO %in% nitrogenase_K_numbers)

print(flexible_nitrogenase_genes)

# Extract flexible genome genes from all Flexible bins
flexible_genes <- RichPangenome %>%
  filter(startsWith(bin_name, "Flexible")) %>%
  pull(EGGNOG_KEGG_KO)

# Check if all nitrogenase genes are in the flexible genome
all_in_flexible <- all(nitrogenase_K_numbers %in% flexible_genes)

# Output the result for flexible genes
if (all_in_flexible) {
  print("All nitrogenase genes are in the flexible genome.")
} else {
  print("Not all nitrogenase genes are in the flexible genome.")
}

# Find common nitrogenase genes in the flexible genome
common_genes <- intersect(nitrogenase_K_numbers, flexible_genes)

# Count how many nitrogenase genes are in the flexible genome
num_common_genes <- length(common_genes)

# Output the result
print(paste(num_common_genes, "out of", length(nitrogenase_K_numbers), "nitrogenase genes are in the flexible genome."))
print("The nitrogenase genes in the flexible genome are:")
print(common_genes)


# Check if nitrogenase genes are marked as singletons
singleton_nitrogenase_genes <- RichPangenome %>%
  filter(bin_name == "Singleton" & EGGNOG_KEGG_KO %in% nitrogenase_K_numbers)

print(singleton_nitrogenase_genes)

# Extract singleton genome genes
singleton_genes <- RichPangenome %>%
  filter(bin_name == "Singleton") %>%
  pull(EGGNOG_KEGG_KO)

# Check if all nitrogenase genes are in the singleton genome
all_in_singleton <- all(nitrogenase_K_numbers %in% singleton_genes)

# Output the result for singleton genes
if (all_in_singleton) {
  print("All nitrogenase genes are in the singleton genome.")
} else {
  print("Not all nitrogenase genes are in the singleton genome.")
}

# Find common nitrogenase genes in the singleton genome
common_genes <- intersect(nitrogenase_K_numbers, singleton_genes)

# Count how many nitrogenase genes are in the singleton genome
num_common_genes <- length(common_genes)

# Output the result
print(paste(num_common_genes, "out of", length(nitrogenase_K_numbers), "nitrogenase genes are in the singleton genome."))
print("The nitrogenase genes in the singleton genome are:")
print(common_genes)

```


#Checking for Photosyntheisis 

```{r}
photosystem_K_numbers <- c(
  # Photosystem II
  "ko:K02703", # psbA - Photosystem II P680 reaction center D1 protein [EC:1.10.3.9]
  "ko:K02704", # psbB - Photosystem II CP47 chlorophyll apoprotein
  "ko:K02705", # psbC - Photosystem II CP43 chlorophyll apoprotein
  "ko:K02706", # psbD - Photosystem II P680 reaction center D2 protein [EC:1.10.3.9]
  "ko:K02707", # psbE - Photosystem II cytochrome b559 subunit alpha
  "ko:K02708", # psbF - Photosystem II cytochrome b559 subunit beta
  "ko:K02709", # psbH - Photosystem II PsbH protein
  "ko:K02710", # psbI - Photosystem II PsbI protein
  "ko:K02711", # psbJ - Photosystem II PsbJ protein
  "ko:K02712", # psbK - Photosystem II PsbK protein
  "ko:K02713", # psbL - Photosystem II PsbL protein
  "ko:K02714", # psbM - Photosystem II PsbM protein
  "ko:K02716", # psbO - Photosystem II oxygen-evolving enhancer protein 1
  "ko:K02717", # psbP - Photosystem II oxygen-evolving enhancer protein 2
  "ko:K08901", # psbQ - Photosystem II oxygen-evolving enhancer protein 3
  "ko:K03541", # psbT - Photosystem II 10kDa protein
  "ko:K03542", # psbS - Photosystem II 22kDa protein
  "ko:K02718", # psbTn - Photosystem II PsbT protein
  "ko:K02719", # psbU - Photosystem II PsbU protein
  "ko:K02720", # psbV - Photosystem II cytochrome c550
  "ko:K02721", # psbW - Photosystem II PsbW protein
  "ko:K02722", # psbX - Photosystem II PsbX protein
  "ko:K02723", # psbY - Photosystem II PsbY protein
  "ko:K02724", # psbZ - Photosystem II PsbZ protein
  "ko:K08902", # psb27 - Photosystem II Psb27 protein
  "ko:K08903", # psb28 - Photosystem II 13kDa protein
  "ko:K08904", # psb28-2 - Photosystem II Psb28-2 protein
  
  # Photosystem I
  "ko:K02689", # psaA - Photosystem I P700 chlorophyll a apoprotein A1 [EC:1.97.1.12]
  "ko:K02690", # psaB - Photosystem I P700 chlorophyll a apoprotein A2 [EC:1.97.1.12]
  "ko:K02691", # psaC - Photosystem I iron-sulfur center [EC:1.97.1.12]
  "ko:K02692", # psaD - Photosystem I subunit II
  "ko:K02693", # psaE - Photosystem I subunit IV
  "ko:K02694", # psaF - Photosystem I subunit III
  "ko:K08905", # psaG - Photosystem I subunit V
  "ko:K02695", # psaH - Photosystem I subunit VI
  "ko:K02696", # psaI - Photosystem I subunit VIII
  "ko:K02697", # psaJ - Photosystem I subunit IX
  "ko:K02698", # psaK - Photosystem I subunit X
  "ko:K02699", # psaL - Photosystem I subunit XI
  "ko:K02700", # psaM - Photosystem I subunit XII
  "ko:K02701", # psaN - Photosystem I subunit PsaN
  "ko:K14332", # psaO - Photosystem I subunit PsaO
  "ko:K02702", # psaX - Photosystem I 4.8kDa protein
  
  # Cytochrome b6-f complex
  "ko:K02634", # petA - Apocytochrome f
  "ko:K02635", # petB - Cytochrome b6
  "ko:K02636", # petC - Cytochrome b6-f complex iron-sulfur subunit [EC:7.1.1.6]
  "ko:K02637", # petD - Cytochrome b6-f complex subunit 4
  "ko:K02640", # petG - Cytochrome b6-f complex subunit 5
  "ko:K02642", # petL - Cytochrome b6-f complex subunit 6
  "ko:K02643", # petM - Cytochrome b6-f complex subunit 7
  "ko:K03689", # petN - Cytochrome b6-f complex subunit 8
  
  # Other proteins involved in photosynthesis
  "ko:K02638", # petE - Plastocyanin
  "ko:K02639", # petF - Ferredoxin
  "ko:K02641", # petH - Ferredoxin--NADP+ reductase [EC:1.18.1.2]
  "ko:K08906", # petJ - Cytochrome c6
  
  # F-type ATPase subunits
  "ko:K02114", # atpC - F-type H+-transporting ATPase subunit epsilon
  "ko:K02112", # atpD - F-type H+/Na+-transporting ATPase subunit beta [EC:7.1.2.2 7.2.2.1]
  "ko:K02115", # atpE - F-type H+-transporting ATPase subunit gamma
  "ko:K02111", # atpA - F-type H+/Na+-transporting ATPase subunit alpha [EC:7.1.2.2 7.2.2.1]
  "ko:K02113", # atpH - F-type H+-transporting ATPase subunit delta
  "ko:K02109", # atpF - F-type H+-transporting ATPase subunit b
  "ko:K02110", # atpG - F-type H+-transporting ATPase subunit c
  "ko:K02108"  # atpI - F-type H+-transporting ATPase subunit a
)


# Check if photosystem K numbers are present in the dataset
photosystem_genes_present <- RichPangenome %>%
  filter(EGGNOG_KEGG_KO %in% photosystem_K_numbers)

# Check if photosystem genes are marked as core
core_photosystem_genes <- RichPangenome %>%
  filter(bin_name == "Core" & EGGNOG_KEGG_KO %in% photosystem_K_numbers)

print(core_photosystem_genes)

# Extract core genome genes
core_genes <- RichPangenome %>%
  filter(bin_name == "Core") %>%
  pull(EGGNOG_KEGG_KO)

# Check if all photosystem genes are in the core genome
all_in_core <- all(photosystem_K_numbers %in% core_genes)

# Output the result for core genes
if (all_in_core) {
  print("All photosystem genes are in the core genome.")
} else {
  print("Not all photosystem genes are in the core genome.")
}

# Extract flexible genome genes from all Flexible bins
flexible_genes <- RichPangenome %>%
  filter(startsWith(bin_name, "Flexible")) %>%
  pull(EGGNOG_KEGG_KO)

# Check if all photosystem genes are in the flexible genome
all_in_flexible <- all(photosystem_K_numbers %in% flexible_genes)

# Output the result for flexible genes
if (all_in_flexible) {
  print("All photosystem genes are in the flexible genome.")
} else {
  print("Not all photosystem genes are in the flexible genome.")
}

# Extract singleton genome genes
singleton_genes <- RichPangenome %>%
  filter(bin_name == "Singleton") %>%
  pull(EGGNOG_KEGG_KO)

# Check if all photosystem genes are in the singleton genome
all_in_singleton <- all(photosystem_K_numbers %in% singleton_genes)

# Output the result for singleton genes
if (all_in_singleton) {
  print("All photosystem genes are in the singleton genome.")
} else {
  print("Not all photosystem genes are in the singleton genome.")
}

# Find common photosystem genes in the core genome
common_core_genes <- intersect(photosystem_K_numbers, core_genes)
print(paste(length(common_core_genes), "out of", length(photosystem_K_numbers), "photosystem genes are in the core genome."))
print("The photosystem genes in the core genome are:")
print(common_core_genes)

# Find common photosystem genes in the flexible genome
common_flexible_genes <- intersect(photosystem_K_numbers, flexible_genes)
print(paste(length(common_flexible_genes), "out of", length(photosystem_K_numbers), "photosystem genes are in the flexible genome."))
print("The photosystem genes in the flexible genome are:")
print(common_flexible_genes)

# Find common photosystem genes in the singleton genome
common_singleton_genes <- intersect(photosystem_K_numbers, singleton_genes)
print(paste(length(common_singleton_genes), "out of", length(photosystem_K_numbers), "photosystem genes are in the singleton genome."))
print("The photosystem genes in the singleton genome are:")
print(common_singleton_genes)

```


#Phycobilismoe K numbers

```{r}
# List of phycobilisome K numbers with descriptions
phycobilisome_K_numbers <- c(
  # Allophycocyanin
  "ko:K02092", # Allophycocyanin alpha subunit
  "ko:K02093", # Allophycocyanin beta subunit
  "ko:K02094", # Phycobilisome core linker protein
  "ko:K02095", # Allophycocyanin-B
  "ko:K02096", # Phycobilisome core-membrane linker protein
  "ko:K02097", # Phycobilisome core component
  
  # Phycocyanin
  "ko:K02284", # Phycocyanin alpha chain
  "ko:K02285", # Phycocyanin beta chain
  "ko:K02286", # Phycocyanin-associated rod linker protein
  "ko:K02287", # Phycocyanin-associated, rod
  "ko:K02288", # Phycocyanobilin lyase subunit alpha [EC:4.4.1.32]
  "ko:K02289", # Phycocyanobilin lyase subunit beta [EC:4.4.1.32]
  "ko:K20713", # R-phycocyanin alpha-cysteine-84 phycourobilin lyase/isomerase [EC:4.4.1.33]
  "ko:K02290", # Phycobilisome rod-core linker protein
  
  # Phycoerythrocyanin
  "ko:K02628", # Phycoerythrocyanin alpha chain
  "ko:K02629", # Phycoerythrocyanin beta chain
  "ko:K02630", # Phycoerythrocyanin-associated rod linker protein
  "ko:K02631", # Phycoerythrocyanin alpha-cysteine-84 phycoviolobilin lyase/isomerase subunit PecE [EC:4.4.1.31]
  "ko:K02632", # Phycoerythrocyanin alpha-cysteine-84 phycoviolobilin lyase/isomerase subunit PecF [EC:4.4.1.31]
  
  # Phycoerythrin
  "ko:K05376", # Phycoerythrin alpha chain
  "ko:K05377", # Phycoerythrin beta chain
  "ko:K05378", # Phycoerythrin-associated linker protein
  "ko:K05379", # Phycoerythrin-associated linker protein
  "ko:K05380", # Phycoerythrin-associated linker protein
  "ko:K05381", # Phycoerythrin-associated linker protein
  "ko:K05382", # Phycoerythrin-associated linker protein
  "ko:K05383", # CpeT protein
  "ko:K05384", # Bilin biosynthesis protein
  "ko:K05385", # Bilin biosynthesis protein
  "ko:K05386", # Bilin biosynthesis protein
  
  # Light-harvesting complex I
  "ko:K08907", # Light-harvesting complex I chlorophyll a/b binding protein 1
  "ko:K08908", # Light-harvesting complex I chlorophyll a/b binding protein 2
  "ko:K08909", # Light-harvesting complex I chlorophyll a/b binding protein 3
  "ko:K08910", # Light-harvesting complex I chlorophyll a/b binding protein 4
  "ko:K08911", # Light-harvesting complex I chlorophyll a/b binding protein 5
  
  # Light-harvesting complex II
  "ko:K08912", # Light-harvesting complex II chlorophyll a/b binding protein 1
  "ko:K08913", # Light-harvesting complex II chlorophyll a/b binding protein 2
  "ko:K08914", # Light-harvesting complex II chlorophyll a/b binding protein 3
  "ko:K08915", # Light-harvesting complex II chlorophyll a/b binding protein 4
  "ko:K08916", # Light-harvesting complex II chlorophyll a/b binding protein 5
  "ko:K08917", # Light-harvesting complex II chlorophyll a/b binding protein 6
  "ko:K14172"  # Light-harvesting complex II chlorophyll a/b binding protein 7
)

# Check if phycobilisome K numbers are present in the dataset
phycobilisome_genes_present <- RichPangenome %>%
  filter(EGGNOG_KEGG_KO %in% phycobilisome_K_numbers)

# Check if phycobilisome genes are marked as core
core_phycobilisome_genes <- RichPangenome %>%
  filter(bin_name == "Core" & EGGNOG_KEGG_KO %in% phycobilisome_K_numbers)

print(core_phycobilisome_genes)

# Extract core genome genes
core_genes <- RichPangenome %>%
  filter(bin_name == "Core") %>%
  pull(EGGNOG_KEGG_KO)

# Check if all phycobilisome genes are in the core genome
all_in_core <- all(phycobilisome_K_numbers %in% core_genes)

# Output the result for core genes
if (all_in_core) {
  print("All phycobilisome genes are in the core genome.")
} else {
  print("Not all phycobilisome genes are in the core genome.")
}

# Extract flexible genome genes from all Flexible bins
flexible_genes <- RichPangenome %>%
  filter(startsWith(bin_name, "Flexible")) %>%
  pull(EGGNOG_KEGG_KO)

# Check if all phycobilisome genes are in the flexible genome
all_in_flexible <- all(phycobilisome_K_numbers %in% flexible_genes)

# Output the result for flexible genes
if (all_in_flexible) {
  print("All phycobilisome genes are in the flexible genome.")
} else {
  print("Not all phycobilisome genes are in the flexible genome.")
}

# Extract singleton genome genes
singleton_genes <- RichPangenome %>%
  filter(bin_name == "Singleton") %>%
  pull(EGGNOG_KEGG_KO)

# Check if all phycobilisome genes are in the singleton genome
all_in_singleton <- all(phycobilisome_K_numbers %in% singleton_genes)

# Output the result for singleton genes
if (all_in_singleton) {
  print("All phycobilisome genes are in the singleton genome.")
} else {
  print("Not all phycobilisome genes are in the singleton genome.")
}

# Find common phycobilisome genes in the core genome
common_core_genes <- intersect(phycobilisome_K_numbers, core_genes)
print(paste(length(common_core_genes), "out of", length(phycobilisome_K_numbers), "phycobilisome genes are in the core genome."))
print("The phycobilisome genes in the core genome are:")
print(common_core_genes)

# Find common phycobilisome genes in the flexible genome
common_flexible_genes <- intersect(phycobilisome_K_numbers, flexible_genes)
print(paste(length(common_flexible_genes), "out of", length(phycobilisome_K_numbers), "phycobilisome genes are in the flexible genome."))
print("The phycobilisome genes in the flexible genome are:")
print(common_flexible_genes)

# Find common phycobilisome genes in the singleton genome
common_singleton_genes <- intersect(phycobilisome_K_numbers, singleton_genes)
print(paste(length(common_singleton_genes), "out of", length(phycobilisome_K_numbers), "phycobilisome genes are in the singleton genome."))
print("The phycobilisome genes in the singleton genome are:")
print(common_singleton_genes)

```


```{r}
# All specific KOs combined
specific_kos <- c(photosystem_K_numbers, phycobilisome_K_numbers)

# Create a presence-absence table for specific KOs
presence_absence_table <- data.frame(
  K0_number = specific_kos,
  Category = factor(rep(c("Photosystem", "Phycobilisome"),
                        times = c(length(photosystem_K_numbers), length(phycobilisome_K_numbers))),
                    levels = c("Photosystem", "Phycobilisome")),
  Core = specific_kos %in% core_genes,
  Flexible = specific_kos %in% flexible_genes,
  Singleton = specific_kos %in% singleton_genes
)

# Convert logicals to integers (0 and 1)
presence_absence_table <- presence_absence_table %>%
  mutate(across(c(Core, Flexible, Singleton), as.integer))

# Convert the data to long format for ggplot2
presence_absence_long <- presence_absence_table %>%
  pivot_longer(cols = -c(K0_number, Category), names_to = "Genome_Group", values_to = "Presence")

# Set the order of the Genome_Group factor
presence_absence_long$Genome_Group <- factor(presence_absence_long$Genome_Group, levels = c("Core", "Flexible", "Singleton"))

# Create the heatmap
photo <- ggplot(presence_absence_long, aes(x = Genome_Group, y = K0_number, fill = factor(Presence))) +
  geom_tile(color = "grey") +
  scale_fill_manual(values = c("0" = "white", "1" = "#499851"), name = "Presence") +
  facet_wrap(~Category, scales = "free_y") +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1),
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    plot.margin = margin(t = 10, r = 100, b = 50, l = 10)
  ) +
  labs(title = "Presence-Absence Heatmap of Photosystem and Phycobilisome KOs", x = "Genome Group", y = "KO Number")

print(photo)

# Save the ggplot figure as a PDF
ggsave("Photosystem_Photosynthesys.pdf", plot = photo, device = "pdf", width = 10, height = 10)
```


#Carbon fixation

```{r}
carbon_fixation_K_numbers <- c(
  # Ribulose-bisphosphate carboxylase
  "ko:K01601", # rbcL - ribulose-bisphosphate carboxylase large chain [EC:4.1.1.39]
  "ko:K01602", # rbcS - ribulose-bisphosphate carboxylase small chain [EC:4.1.1.39]
  
  # Phosphoglycerate kinase
  "ko:K00927", # pgk - phosphoglycerate kinase [EC:2.7.2.3]
  
  # Glyceraldehyde 3-phosphate dehydrogenase
  "ko:K00134", # gapA - glyceraldehyde 3-phosphate dehydrogenase (phosphorylating) [EC:1.2.1.12]
  "ko:K05298", # gapN - glyceraldehyde-3-phosphate dehydrogenase (NADP+) (phosphorylating) [EC:1.2.1.13]
  "ko:K00150", # gapC - glyceraldehyde-3-phosphate dehydrogenase (NAD(P)+) (phosphorylating) [EC:1.2.1.59]
  
  # Fructose-bisphosphate aldolase
  "ko:K01623", # fbaA - fructose-bisphosphate aldolase, class I [EC:4.1.2.13]
  "ko:K11645", # fbaB - fructose-bisphosphate aldolase, class I [EC:4.1.2.13]
  "ko:K01624", # fbaC - fructose-bisphosphate aldolase, class II [EC:4.1.2.13]
  
  # Fructose-1,6-bisphosphatase
  "ko:K03841", # fbpA - fructose-1,6-bisphosphatase I [EC:3.1.3.11]
  "ko:K02446", # fbpB - fructose-1,6-bisphosphatase II [EC:3.1.3.11]
  "ko:K11532", # fbpC - fructose-1,6-bisphosphatase II / sedoheptulose-1,7-bisphosphatase [EC:3.1.3.11 3.1.3.37]
  "ko:K01086", # fbpD - fructose-1,6-bisphosphatase I / sedoheptulose-1,7-bisphosphatase [EC:3.1.3.11 3.1.3.37]
  "ko:K04041", # fbpE - fructose-1,6-bisphosphatase III [EC:3.1.3.11]
  
  # Transketolase
  "ko:K00615", # tktA - transketolase [EC:2.2.1.1]
  
  # Sedoheptulose-bisphosphatase
  "ko:K01100", # sbp - sedoheptulose-bisphosphatase [EC:3.1.3.37]
  
  # Ribose 5-phosphate isomerase
  "ko:K01807", # rpiA - ribose 5-phosphate isomerase A [EC:5.3.1.6]
  "ko:K01808", # rpiB - ribose 5-phosphate isomerase B [EC:5.3.1.6]
  
  # Phosphoribulokinase
  "ko:K00855", # prkB - phosphoribulokinase [EC:2.7.1.19]
  
  # Triosephosphate isomerase (TIM)
  "ko:K01803", # tpiA - triosephosphate isomerase (TIM) [EC:5.3.1.1]
  
  # Ribulose-phosphate 3-epimerase
  "ko:K01783", # rpe - ribulose-phosphate 3-epimerase [EC:5.1.3.1]
  
  # Transaldolase
  "ko:K00616", # talA - transaldolase [EC:2.2.1.2]
  "ko:K13810", # talB - transaldolase / glucose-6-phosphate isomerase [EC:2.2.1.2 5.3.1.9]
  
  # Phosphoenolpyruvate carboxylase
  "ko:K01595", # ppc - phosphoenolpyruvate carboxylase [EC:4.1.1.31]
  
  # Aspartate aminotransferase
  "ko:K14454", # got1 - aspartate aminotransferase, cytoplasmic [EC:2.6.1.1]
  "ko:K14455", # got2 - aspartate aminotransferase, mitochondrial [EC:2.6.1.1]
  
  # Phosphoenolpyruvate carboxykinase (ATP)
  "ko:K01610", # pckA - phosphoenolpyruvate carboxykinase (ATP) [EC:4.1.1.49]
  
  # Alanine transaminase
  "ko:K00814", # altA - alanine transaminase [EC:2.6.1.2]
  
  # Glutamate-glyoxylate aminotransferase
  "ko:K14272", # ggtA - glutamate--glyoxylate aminotransferase [EC:2.6.1.4 2.6.1.2 2.6.1.44]
  
  # Pyruvate, orthophosphate dikinase
  "ko:K01006", # pdkA - pyruvate, orthophosphate dikinase [EC:2.7.9.1]
  
  # Malate dehydrogenase
  "ko:K00029", # mdh1 - malate dehydrogenase (oxaloacetate-decarboxylating)(NADP+) [EC:1.1.1.40]
  "ko:K00051", # mdh2 - malate dehydrogenase (NADP+) [EC:1.1.1.82]
  "ko:K00025", # mdh3 - malate dehydrogenase [EC:1.1.1.37]
  "ko:K00026", # mdh4 - malate dehydrogenase [EC:1.1.1.37]
  "ko:K00024", # mdh5 - malate dehydrogenase [EC:1.1.1.37]
  "ko:K00028", # meA - malate dehydrogenase (decarboxylating) [EC:1.1.1.39]
)

```




```{r}
# List of KO numbers to check

# Check if combined K numbers are present in the dataset
combined_genes_present <- RichPangenome %>%
  filter(EGGNOG_KEGG_KO %in% combined_K_numbers)

# Check if combined genes are marked as core
core_combined_genes <- RichPangenome %>%
  filter(bin_name == "Core" & EGGNOG_KEGG_KO %in% combined_K_numbers)

print(core_combined_genes)

# Extract core genome genes
core_genes <- RichPangenome %>%
  filter(bin_name == "Core") %>%
  pull(EGGNOG_KEGG_KO)

# Check if all combined genes are in the core genome
all_in_core <- all(combined_K_numbers %in% core_genes)

# Output the result for core genes
if (all_in_core) {
  print("All combined genes are in the core genome.")
} else {
  print("Not all combined genes are in the core genome.")
}

# Extract flexible genome genes from all Flexible bins
flexible_genes <- RichPangenome %>%
  filter(startsWith(bin_name, "Flexible")) %>%
  pull(EGGNOG_KEGG_KO)

# Check if all combined genes are in the flexible genome
all_in_flexible <- all(combined_K_numbers %in% flexible_genes)

# Output the result for flexible genes
if (all_in_flexible) {
  print("All combined genes are in the flexible genome.")
} else {
  print("Not all combined genes are in the flexible genome.")
}

# Extract singleton genome genes
singleton_genes <- RichPangenome %>%
  filter(bin_name == "Singleton") %>%
  pull(EGGNOG_KEGG_KO)

# Check if all combined genes are in the singleton genome
all_in_singleton <- all(combined_K_numbers %in% singleton_genes)

# Output the result for singleton genes
if (all_in_singleton) {
  print("All combined genes are in the singleton genome.")
} else {
  print("Not all combined genes are in the singleton genome.")
}

# Find common combined genes in the core genome
common_core_genes <- intersect(combined_K_numbers, core_genes)
print(paste(length(common_core_genes), "out of", length(combined_K_numbers), "combined genes are in the core genome."))
print("The combined genes in the core genome are:")
print(common_core_genes)

# Find common combined genes in the flexible genome
common_flexible_genes <- intersect(combined_K_numbers, flexible_genes)
print(paste(length(common_flexible_genes), "out of", length(combined_K_numbers), "combined genes are in the flexible genome."))
print("The combined genes in the flexible genome are:")
print(common_flexible_genes)

# Find common combined genes in the singleton genome
common_singleton_genes <- intersect(combined_K_numbers, singleton_genes)
print(paste(length(common_singleton_genes), "out of", length(combined_K_numbers), "combined genes are in the singleton genome."))
print("The combined genes in the singleton genome are:")
print(common_singleton_genes)

```

